### ISSUE TITLE
WeakReferenceList behavior maybe no safe

### ISSUE THREAD

**Issue by lindexi** [dotnet/wpf#3012]
This is the code.

https://github.com/dotnet/wpf/blob/87115e5b0efade5ec32379264e9909d11b01a9e4/src/Microsoft.DotNet.Wpf/src/Shared/MS/Internal/WeakReferenceList.cs#L187-L189

And as we can see, the code check the WeakReference is alive and then get the Target. But the GC will collect the object before we get the Target and after we check the IsAlive.

And I think we should use the TryGetTarget to get the value.

Thank you.

**Comment by weltkante** [comment]
Yes there is a race condition between the two, you need to check obj for null if you use the old API. TryGetTarget requires switching to the newer generic WeakReference.
